blog/2024-05-08-prefixes.mdx

```diff
@@ -0,0 +1,132 @@
+occurred when the `conda-env` project was [merged][conda-env-pr] with the
+`conda` project, elevating `conda` from a mere package manager to a
+comprehensive environment manager. This pivotal transformation bestowed upon
+`conda` the capability to create and manage separate environments, fostering
+modularity and flexibility in development workflows.
+
+Delving into `conda`'s Git history via `git blame` provides invaluable insights
+into the timeline of the prefix system evolution. By tracing the commit history,
+we unearth pivotal moments that introduced and refined the various prefixes
+integral to `conda`'s architecture. The history is also revealing in how
+non-linear the development of the prefix system has been, explaining many of the
+complexities of trying to wrestle with the prefixes or trying to find ways to
+simplify them.
+
+1. `conda.base.context.context.conda_prefix`
+
+   Introduced in `conda 4.2.0` (see https://github.com/conda/conda/commit/6e730cc)
+   as part of the private environment effort which has since been abandoned and
+   deprecated.
+
+   The `conda_prefix` lies at the heart of the prefix system, denoting the
+   installation directory of `conda` itself. Typically, this aligns with the
+   location of the `base` environment and is equivalent to Python's `sys.prefix`
+   value.
+
+   :::info
+
+   Use this anytime you need to interact with the `conda` installation itself.
+
+   :::
+
+2. `conda.base.context.context.root_prefix`
+
+   > Previously `conda.config.root_dir` → `conda.config.ROOT_DIR`.
+
+   Introduced in `conda 1.1.0` (see https://github.com/conda/conda/commit/d2b370a)
+   and was originally intended for testing purposes only.
+
+   The `root_prefix` serves as a [configurable value][condarc], most easily
+   configurable in the user's `~/.condarc` file. In the absence of user-defined
+   preferences, it defaults to the `conda_prefix`.
+
+   :::warning
+
+   Avoid using this. This is a contender for deprecation.
+
+   :::
+
+3. `conda.base.context.context.active_prefix`
+
+   Introduced in `conda 4.4.0` (see https://github.com/conda/conda/commit/cbbe81f)
+   as part of cleaning up the `conda info` output.
+
+   Reflecting the currently active environment, the `active_prefix` is determined
+   by the `$CONDA_PREFIX` environment variable which is set by the `conda activate`
+   and `conda deactivate` commands.
+
+   :::warning
+
+   Avoid using this. Treat this as an internal value.
+
+   :::
+
+4. `conda.base.context.context.default_prefix`
+
+   > Previously `conda.config.default_prefix` → `conda.config.DEFAULT_ENV_PREFIX` →
+   > `conda.config.default_environment` → `conda.anaconda.default_environment`.
+
+   Introduced in `conda 1.1.0` (see https://github.com/conda/conda/commit/acd8144)
+   as part of a larger configuration overhaul.
+
+   The `default_prefix` effectively aggregates `root_prefix` and `active_prefix`,
+   embodying the default environment in the absence of CLI overrides.
```

Comment:
I'm not totally sure what the last part of this sentence this means.

Reply:
When we specify `--name` or `--prefix` on the CLI we are effectively overriding the `default_prefix`, I agree the wording here is not clear, any suggestions would be great otherwise I'll wordsmith some more.

Reply:
Realized this was too verbose. Trimmed.

Reply:
It's mainly the "embodying the default environment" bit that was tripping me up. 

So it aggregates (or combines together) `root_prefix` and `active_prefix`. `root_prefix` is configurable, but defaults to `conda_prefix`, which is conda's install directory, while `active_prefix` is based on what environment you have active.

Yeah, I guess I'm not sure what the default environment is in this context.

Reply:
Your right, that didn't describe anything concretely. Corrected.